### DISCUSSION TITLE
Can't reach database server, error_code":"P1001

### DISCUSSION THREAD

**Discussion by kavenugop** [prisma/prisma#22755] [Q&A]
### Question

Hello I am getting the same P1001 error  multiple times when making prisma connection for a python process that run a lot of CRUD operations. Though the database is up and it connects and runs a few CRUD operations, it fails eventually during the process execution and never recovers. I did put a connection retry logic but the exception is never raised for me to retry connection. Appreciate any guidance.  

```
{"is_panic":false,"message":"Can't reach database server at <postgres server at neon.tech>:5432\n\nPlease make sure your database server is running at <server name>:5432.","meta":{"database_host":"postgres server at neon.tech","database_port":5432},"error_code":"P1001"}
```

```
generator client {
  provider             = "prisma-client-py"
  recursive_type_depth = "5"
}

datasource db {
  provider = "postgres"
  url      = "postgres://<postgres connection url>>?schema=<schema>&connect_timeout=30"
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by janpio** [comment]
The repository for the (community) Python Client is over here: https://github.com/RobertCraigie/prisma-client-py

**Comment by kavenugop** [comment]
@janpio - I check with Prisma Python / Robert over discord and he suggested me to contact Prisma core. Below was his comment - 

"I suspect this is an issue with the underlying core Prisma implementation.

I'm unfortunately not very familiar with this aspect of Prisma so you'd have better luck opening a discussion on the main prisma repo or asking a question in their slack"

**Reply by janpio** [reply]
Ok. Generally, if Prisma tells you it can not connect, then exactly that is the case. Reasons can be many, but nothing we can really debug externally - something leads to Prisma not being able to connect to the database in the defined timeout.

**Reply by kavenugop** [reply]
@janpio  Below are the details of the environment. It would be helpful to understand if there are alternative fixes like connection retries, connection times out etc. I tried the connection time out on the postgre side but did not help. 

One another issue is that the process  does not "raise" an error it just hangs and dies... which means I am not able to catch it and take alterntive actions. 

OS: Windows
Database: Postgres
Node.js version: 18.18.0
Run prisma -v to see your Prisma version and paste it
prisma                : 5.4.2
@prisma/client        : Not found
Current platform      : windows
Query Engine (Binary) : query-engine ac9d7041ed77bcc8a8dbd2ab6616b39013829574 (at ..\..\..\..\..\Users\kisho\.cache\prisma-python\binaries\5.4.2\ac9d7041ed77bcc8a8dbd2ab6616b39013829574\node_modules\@prisma\engines\query-engine-windows.exe)
Schema Engine         : schema-engine-cli ac9d7041ed77bcc8a8dbd2ab6616b39013829574 (at ..\..\..\..\..\Users\kisho\.cache\prisma-python\binaries\5.4.2\ac9d7041ed77bcc8a8dbd2ab6616b39013829574\node_modules\@prisma\engines\schema-engine-windows.exe)
Schema Wasm           : @prisma/prisma-schema-wasm 5.4.1-2.ac9d7041ed77bcc8a8dbd2ab6616b39013829574
Default Engines Hash  : ac9d7041ed77bcc8a8dbd2ab6616b39013829574
Studio                : 0.494.0

**Reply by janpio** [reply]
The only signal that Prisma has that the database server is not reachable is that the connection was not successfully created after the timeout. So you can either increase the timeout, or try again. But usually this should not be such a big problem.

Does it work if you use the Javascript/Node Prisma instead of the Python one?

**Reply by kavenugop** [reply]
I have not tried on the javascript side however I changed the cloud provider and it does work now. The main concern I have is the process getting hung on the connection instead of raising the error back to the client. I wanted to clarify if this is an issue with the core prisma implementation or do you think its related to the python implementation?

**Reply by janpio** [reply]
In JS this would definitely be raised in the app after the timeout duration - so if that is _not_ happening in Python, then this is specific to Python and needs to be reported (and hopefully fixed) there.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
